File: pkg/cli/cmd/find/lexer.go
/* Copyright 2025 Dnote Authors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package find

import (
	"regexp"
)

var newLineReg = regexp.MustCompile(`\r?\n`)

const (
	// tokenKindChar represents utf-8 character
	tokenKindChar = iota
	// tokenKindHLBegin represents a beginning of a highlighted section
	tokenKindHLBegin
	// tokenKindHLEnd represents an end of a highlighted section
	tokenKindHLEnd
	// tokenKindEOL represents an end of line
	tokenKindEOL
)

type token struct {
	Value byte
	Kind  int
}

// getNextIdx validates that the given index is within the range of the given string.
// If so, it returns the given index. Otherwise it returns -1.
func getNextIdx(candidate int, s string) int {
	if candidate <= len(s)-1 {
		return candidate
	}

	return -1
}

// scanToken scans the given string for a token at the given index. It returns
// a token and the next index to look for a token. If the given string is exhausted,
// the next index will be -1.
func scanToken(idx int, s string) (token, int) {
	if s[idx] == '<' {
		if len(s)-idx >= 9 {
			lookahead := 9
			candidate := s[idx : idx+lookahead]

			if candidate == "<dnotehl>" {
				nextIdx := getNextIdx(idx+lookahead, s)
				return token{Kind: tokenKindHLBegin}, nextIdx
			}
		}

		if len(s)-idx >= 10 {
			lookahead := 10
			candidate := s[idx : idx+lookahead]

			if candidate == "</dnotehl>" {
				nextIdx := getNextIdx(idx+lookahead, s)
				return token{Kind: tokenKindHLEnd}, nextIdx
			}
		}
	}

	nextIdx := getNextIdx(idx+1, s)

	return token{Value: s[idx], Kind: tokenKindChar}, nextIdx
}

// tokenize lexically analyzes the given matched snippet from a full text search
// and builds a slice of tokens
func tokenize(s string) []token {
	var ret []token

	idx := 0
	for idx != -1 {
		var tok token
		tok, idx = scanToken(idx, s)

		ret = append(ret, tok)
	}

	ret = append(ret, token{Kind: tokenKindEOL})

	return ret
}

File: web/src/libs/fts/lexer.ts
/* Copyright (C) 2019 Monomax Software Pty Ltd
 *
 * This file is part of Dnote.
 *
 * Dnote is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * Dnote is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with Dnote.  If not, see <https://www.gnu.org/licenses/>.
 */

export enum TokenKind {
  char,
  hlBegin,
  hlEnd,
  eol
}

export interface Token {
  kind: TokenKind;
  value?: string;
}

interface ScanTokenResult {
  tok: Token;
  nextIdx: number;
}

// getNextIdx validates that the given index is within the range of the given string.
// If so, it returns the given index. Otherwise it returns -1.
function getNextIdx(candidate: number, s: string): number {
  if (candidate <= s.length - 1) {
    return candidate;
  }

  return -1;
}

// scanToken scans the given string for a token at the given index. It returns
// a token and the next index to look for a token. If the given string is exhausted,
// the next index will be -1.
export function scanToken(idx: number, s: string): ScanTokenResult {
  if (s[idx] === '<') {
    if (s.length - idx >= 9) {
      const lookahead = 9;
      const candidate = s.substring(idx, idx + lookahead);

      if (candidate === '<dnotehl>') {
        return {
          tok: {
            kind: TokenKind.hlBegin
          },
          nextIdx: getNextIdx(idx + lookahead, s)
        };
      }
    }

    if (s.length - idx >= 10) {
      const lookahead = 10;
      const candidate = s.substring(idx, idx + lookahead);

      if (candidate === '</dnotehl>') {
        return {
          tok: {
            kind: TokenKind.hlEnd
          },
          nextIdx: getNextIdx(idx + lookahead, s)
        };
      }
    }
  }

  const nextIdx = getNextIdx(idx + 1, s);
  return {
    tok: {
      value: s[idx],
      kind: TokenKind.char
    },
    nextIdx
  };
}

// tokenize lexically analyzes the given matched snippet from a full text search
// and builds a slice of tokens
export function tokenize(s: string): Token[] {
  const ret: Token[] = [];

  let idx = 0;
  while (idx !== -1) {
    const { tok, nextIdx } = scanToken(idx, s);

    idx = nextIdx;
    ret.push(tok);
  }

  ret.push({ kind: TokenKind.eol });

  return ret;
}
